### ISSUE TITLE
Overloading of interface default methods not working

### ISSUE THREAD

**Issue by Ablu** [cincheo/jsweet#289]
This code:

~~~java
	interface A {
		public default void test(String s) {}
		public default void test(int i) {}
	}

	class B implements A {
	}
~~~

Generates something like:
~~~ts
        export interface A {
            test(s? : any) : any;
        }

        export class B implements API.A {
            public __parent: any;
            public test(i : number) {
            }
            public test(s : string) {
            }
            constructor(__parent: any) {
                this.__parent = __parent;
            }
        }
~~~

`B` then does not implement the interface, and also illegally overloads a function.

**Comment by renaudpawlak** [comment]
Thanks for reporting that bug. I need to fully re-implement default methods anyway. The way it is done is wrong (probably lack of sleep ;))

I'll try to plan it for v2...

**Comment by joslarson** [comment]
@renaudpawlak I see you have this planned for v2.1. I'm on 2.0-rc1 and I'm running into what I assume is the same issue, but I want to make sure.

Am I correct to assume that the the following Java interface...

```java
public interface Foo {
      void bar();

      void bar(String name);

      void bar(String name, int numSends);

      void bar(String id, String name, int size, String mode);
}
```
...should transpile to the following typescript?

```typescript
interface Foo {
      bar(): void;
      bar(name: string): void;
      bar(name: string, size: number): void;
      bar(id: string, name: string, size: number, mode: string): void;
}
```

It currently transpiles to this:

```typescript
interface Foo {
    bar(id? : any, name? : any, size? : any, mode? : any) : any;
}
```

**Comment by renaudpawlak** [comment]
Nope, that a different issue, but it is not exactly a bug I'd say (unless it raises compilation errors). It would be more an improvement in the TypeScript generated code in case of overloading. 

What you see is just the default behaviour when JSweet meets overloading. It take the longest method and relax parameter typings, so that you can call any overloaded method.

Of course, this could be improved, especially for interfaces, were overloading is possible in TypeScript. Also, parameter naming should be improved... I have identified this issue before... It is part of a global improvement with method overloading, but it is not planned for v2, since most use cases I have met so far can work "okay" without it.

Thank you for reporting this problem anyway. I hope it's fine for you to wait until 2.1.

**Comment by joslarson** [comment]
@renaudpawlak thanks for the reply, and yes, of course I can wait :) I really appreciate your work on this. Would you like me to create a separate issue to track this?

Also you mentioned overloading being possible in TypeScript interfaces, and that syntax is also available in class method declarations like so:

```typescript
class Foo {
    bar(): void;
    bar(name: string): void;
    bar(name: string, size: number): void;
    bar(id: string, name: string, size: number, mode: string): void;
    bar(...args: any[]) {  // this is just the implementation signature and is not exposed externally as a valid call signature
        // ...implementation
    }
}
```

You likely already know this, just wanted to make sure.

**Comment by renaudpawlak** [comment]
@joslarson Thank you for sharing. I don't have always time to catch-up with TypeScript features... I am totally sharing your concerns. I think better supporting overloading from an API perspective should be high priority for next release. Note that it should not be that hard... It is related to the issue of generating sound JSDoc comments in case of overloading... 

Yes, since it is really not at all the same issue as default methods (which is a less crucial issue I believe), it would be good to have it in a separate issue. Don't hesitate to open one. Cheers.

**Comment by joslarson** [comment]
@renaudpawlak thanks, I've created a separate issue as requested: #328 

@Ablu sorry for hijacking your issue

**Comment by PashaTurok** [comment]
Interface in TypeScript can not have implementation as interfaces don't exists in JavaScript code. So, I see here two ways: 
1) create interface and one abstract class that implements this interface and has implementation for default methods 
2) use abstract class instead of interface + mixins as described [here](https://stackoverflow.com/a/55784652/5057736). Both variants are not good 1) breaks API 2) mixins is executed for many classes and there are two cycles. 

Maybe there are other solutions?
